### PR TITLE
prototype implementation of core client builder functionality

### PR DESCRIPTION
**Pull request by jpower432**
## Changes from this PR

This PR adds initial building libraries and reference implementation for building capabilities.

## TODO (Follow-on PR)
Add structured logging solution for cli package (logrus?)
CI for automated unit testing

/kind design

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,3 +10,16 @@ embedded in UOR artifacts.
 
 To learn more about Universal Runtime visit the [UOR-Framework](https://github.com/uor-framework/uor-framework) repository.
 
+## Basic Usage
+
+### Template content in a directory without pushing 
+```
+# The default workspace is "client-workspace" in the current working directory
+client build <directory> --output my-workspace
```
Comment (guymguym) [inline README.md:18]: I think this Usage section is a great idea, but it's too hard to understand - maybe we can add a longer flow, starting from how to build the client from source, how to start a local registry, and how to run the client a few times with different options. I think the initial audience of the README are the first contributors to the project, so we want as many people who see this first time to be able to get this working locally so they can hack with it a bit or try to integrate it into another client.

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,38 @@
+GO := go
+
+GO_BUILD_PACKAGES := ./cmd/...
+GO_BUILD_BINDIR :=./bin
+
+all: clean vendor test-unit build
```
Comment (guymguym) [inline Makefile:6]: always clean? I tend to set default targets to do build-over for development speed.
It has its downsides, but nothing that `make clean` can't solve :)

File: cli/build.go [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+	templateBuilder := builder.NewBuilder(userSpace)
+	renderSpace, err := workspace.NewLocalWorkspace(o.Output)
+	if err != nil {
+		return err
+	}
+	if err := templateBuilder.Run(ctx, g, renderSpace); err != nil {
+		return fmt.Errorf("error building content: %v", err)
+	}
+
+	if o.Push {
```
Comment (guymguym) [inline cli/build.go:188]: Do you think we want to move this to a function so we can also use it to push a pre-built artifact? I don't even know if we have such an intermediate output of the artifact but just thinking out load here.

File: builder/graph/graph.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+func NewNode(name string) *Node {
+	return &Node{
+		Name:  name,
+		Nodes: map[string]*Node{},
+		Links: map[string]interface{}{},
+	}
+}
+
+// Graph defines a collection of Nodes.
+type Graph struct {
```
Comment (guymguym) [inline builder/graph/graph.go:32]: Maybe worth adding to the code comment how this Graph struct is used in the sense of does it represent a single OCI Artifact, or is it used to load multiple artifacts? The graph defines connections between keys, but it does not contain any other information per key, such as it's location, content, size, etc which is defined in ... (?)

**Review by guymguym — COMMENTED** [review]
LVGTM 👍 
I tried to be useful and give some feedback but I couldn't go over all the code, and just posted few minor comments above, but since this PR is very foundational, feel free to skip comments that don't seem important for now. Looks great! Thanks!

**Comment by afflom** [comment]
I've really enjoyed playing with this client and I was already able to integrate it into a couple UOR demos. You really went above and beyond for this first PR. 

Behaviorally, there are two issues that I have found:

1. oras pull will write the retrieved artifacts to the same directory structure that they were published from. The uor client uses absolute paths in publishing, which require the retrieved artifacts to be written to unexpected locations. 
2. Hyphens are disallowed by the client's templating scheme. That is inconvenient. 

Thank you for this PR. Your code is very fun to hack on.

File: README.md (hunk 1 above)
Comment (jpower432) [reply README.md:18]: I agree this README could jazzed up quite a bit to attract more people and the Usage section kind of comes out of nowhere. Great comment. I will tackle this on a follow-on PR to give it some thought.

File: Makefile (hunk 2 above)
Comment (jpower432) [reply Makefile:6]: Fair :). I can move this down in the list so it's not the default target.

File: cli/build.go (hunk 3 above)
Comment (jpower432) [reply cli/build.go:188]: I played around with the idea of separating build and push into different sub-commands. I think the workflow you are suggestion here could help us support clients like `dnf` or `yum` which would need an intermediary step between build and push. I will take that as an action on a subsequent PR.

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,38 @@
+GO := go
+
+GO_BUILD_PACKAGES := ./cmd/...
+GO_BUILD_BINDIR :=./bin
+
+build:
+	mkdir -p ${GO_BUILD_BINDIR}
+	GOOS=linux GOARCH=amd64 $(GO) build -o $(GO_BUILD_BINDIR)/client $(GO_BUILD_PACKAGES)
```
Comment (sabre1041) [inline Makefile:8]: Can we either disable since platform build or enable cross platform build?

File: README.md [hunk 6]
```diff
@@ -10,3 +10,16 @@ embedded in UOR artifacts.
 
 To learn more about Universal Runtime visit the [UOR-Framework](https://github.com/uor-framework/uor-framework) repository.
```
Comment (sabre1041) [inline README.md:11]: We should point them at the end user site, and not the repository containing the site content

**Review by sabre1041 — CHANGES_REQUESTED** [review]
Great first implementation. Small request then we can push this through

**Review by sabre1041 — APPROVED** [review]
LGTM

**Review by guymguym — APPROVED** [review]

**Pull request merged**